In [1]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings("ignore")

In [3]:
df = pd.read_csv("dataset\\uk_accident_data.csv")

In [25]:
df

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    05/06/2019  51.506187   
1       200701BS65737           Serious    02/07/2019  51.495029   
2       200701BS66127           Serious    26/08/2019  51.517715   
3       200701BS66128           Serious    16/08/2019  51.495478   
4       200701BS66837            Slight    03/09/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18/02/2022  57.374005   
660675  201091NM01881            Slight    21/02/2022  57.232273   
660676  201091NM01935            Slight    23/02/2022  57.585044   
660677  201091NM01964           Serious    23/02/2022  57.214898   
660678  201091NM02142           Serious    28/02/2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2       Single carriageway               Urban     Fine no high winds   
3       Single carriageway               Urban     Fine no high winds   
4       Single carriageway               Urban     Fine no high winds   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [7]:
df.describe().T

count       mean       std        min        25%  \
Latitude              660654.0  52.553866  1.406922  49.914430  51.490690   
Longitude             660653.0  -1.431210  1.383330  -7.516225  -2.332291   
Number_of_Casualties  660679.0   1.357040  0.824847   1.000000   1.000000   
Number_of_Vehicles    660679.0   1.831255  0.715269   1.000000   1.000000   

                            50%        75%        max  
Latitude              52.315641  53.453452  60.757544  
Longitude             -1.411667  -0.232869   1.762010  
Number_of_Casualties   1.000000   1.000000  68.000000  
Number_of_Vehicles     2.000000   2.000000  32.000000

In [8]:
df.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [13]:
df['Latitude']=df['Latitude'].fillna(df['Latitude'].mean())
df['Longitude']=df['Longitude'].fillna(df['Longitude'].mean())
df['Road_Surface_Conditions']=df['Road_Surface_Conditions'].fillna(df['Road_Surface_Conditions'].mode()[0])
df['Road_Type']=df['Road_Type'].fillna(df['Road_Type'].mode()[0])
df['Urban_or_Rural_Area']=df['Urban_or_Rural_Area'].fillna(df['Urban_or_Rural_Area'].mode()[0])
df['Weather_Conditions']=df['Weather_Conditions'].fillna(df['Weather_Conditions'].mode()[0])

In [14]:
df.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype   
---  ------                   --------------   -----   
 0   Index                    660679 non-null  category
 1   Accident_Severity        660679 non-null  object  
 2   Accident Date            660679 non-null  object  
 3   Latitude                 660679 non-null  float64 
 4   Light_Conditions         660679 non-null  object  
 5   District Area            660679 non-null  object  
 6   Longitude                660679 non-null  float64 
 7   Number_of_Casualties     660679 non-null  int64   
 8   Number_of_Vehicles       660679 non-null  int64   
 9   Road_Surface_Conditions  660679 non-null  object  
 10  Road_Type                660679 non-null  object  
 11  Urban_or_Rural_Area      660679 non-null  object  
 12  Weather_Conditions       660679 non-null  object  
 13  Vehicle_Type             660679 non-null  ob

In [18]:
df['Index']=df['Index'].astype('category')
df['Accident_Severity']=df['Accident_Severity'].astype('category')
df['Light_Conditions']=df['Light_Conditions'].astype('category')
df['District Area']=df['District Area'].astype('category')
df['Road_Surface_Conditions']=df['Road_Surface_Conditions'].astype('category')
df['Road_Type']=df['Road_Type'].astype('category')
df['Urban_or_Rural_Area']=df['Urban_or_Rural_Area'].astype('category')
df['Weather_Conditions']=df['Weather_Conditions'].astype('category')
df['Vehicle_Type']=df['Vehicle_Type'].astype('category')

In [41]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 15 columns):
 #   Column                   Non-Null Count   Dtype   
---  ------                   --------------   -----   
 0   Index                    660679 non-null  category
 1   Accident_Severity        660679 non-null  category
 2   Accident Date            660679 non-null  object  
 3   Latitude                 660679 non-null  float64 
 4   Light_Conditions         660679 non-null  category
 5   District Area            660679 non-null  category
 6   Longitude                660679 non-null  float64 
 7   Number_of_Casualties     660679 non-null  int64   
 8   Number_of_Vehicles       660679 non-null  int64   
 9   Road_Surface_Conditions  660679 non-null  category
 10  Road_Type                660679 non-null  category
 11  Urban_or_Rural_Area      660679 non-null  category
 12  Weather_Conditions       660679 non-null  category
 13  Vehicle_Type             660679 non-null  ca

<h1>Question1: What proportion of accidents are fatal, serious, or slight?</h1>
<hr>

In [5]:
severity_counts = df['Accident_Severity'].value_counts(normalize=True) * 100

In [6]:
severity_counts

Accident_Severity
Slight     85.336601
Serious    13.352475
Fatal       1.310924
Name: proportion, dtype: float64

<h3>INSIGHT: Fatal accidents are 1.310924, serious around 13.352475, and slight accidents 85.336601. This reflects how most collisions occur at low speeds in urban settings. </h3>
<hr>

<h1>Question2: Which regions (urban vs rural) have the highest accidents?</h1>
<hr>

In [23]:
regions = df['Urban_or_Rural_Area'].value_counts()

In [24]:
regions

Urban_or_Rural_Area
Urban          421678
Rural          238990
Unallocated        11
Name: count, dtype: int64

<h3>INSIGHT: Urban happens to have the highest accidents with 421678, while Rural has 238990.</h3>
<hr>

<h1>Question3: Which regions (urban vs rural) have the highest fatal accident rates?</h1>
<hr>

In [21]:
regions_fatal = df[df['Accident_Severity'] == 'Fatal']['Urban_or_Rural_Area'].value_counts()

In [22]:
regions_fatal

Urban_or_Rural_Area
Rural          5601
Urban          3060
Unallocated       0
Name: count, dtype: int64

<h3> INSIGHT: Rural roads often have fewer accidents but a higher share of fatalities with 5601. High-speed collisions and delayed emergency response times increase severity outside cities.</h3>
<hr>

<h1>Question4: Which month has the highest accident frequency?</h1>
<hr>

In [42]:
df["Accident Date"] = pd.to_datetime(df["Accident Date"], errors="coerce")

In [43]:
df["Month"] = df["Accident Date"].dt.month
df["Month_Name"] = df["Accident Date"].dt.month_name()

In [44]:
accidents_by_month = df["Month_Name"].value_counts().reindex(
    ["January","February","March","April","May","June",
     "July","August","September","October","November","December"]
)

In [45]:
accidents_by_month

Month_Name
January      22606
February     21815
March        21540
April        21699
May          22409
June         21974
July         21431
August       21914
September    22252
October      22328
November     22503
December     22536
Name: count, dtype: int64

<h3>INSIGHT: Accident totals usually peak in January. Summer often sees more accidents due to higher travel volumes. Winter may reduce total accidents but increase severity from ice.</h3>
<hr>

<h1>Question5: How many accidents occur per year?</h1>
<hr>

In [51]:
df['Accident Date'] = pd.to_datetime(df['Accident Date'])
accidents_per_year = df['Accident Date'].dt.year.value_counts().sort_index()
accidents_per_year

Accident Date
2019.0    71867
2020.0    70163
2021.0    66172
2022.0    56805
Name: count, dtype: int64

<h3>INSIGHT: In 2019 the number of accidents reach 71867.However, The number of accidents are varies from year to year. These fluctuations can reflect broader factors such as road safety campaigns, traffic enforcement intensity, or overall changes in travel behavior.</h3>
<hr>

<h1>Question6: What Light Condition or time of day has the most accidents?</h1>
<hr>

In [48]:
df['Light_Conditions'].value_counts()

Light_Conditions
Daylight                       484880
Darkness - lights lit          129335
Darkness - no lighting          37437
Darkness - lighting unknown      6484
Darkness - lights unlit          2543
Name: count, dtype: int64

<h3>INSIGHT:Daylight has 484880 which is the most accidents.Rush hours (morning and evening) see spikes in crashes However, late-night accidents are more likely to be severe.</h3>
<hr>

<h1>Question7: Which road types see the most accidents?</h1>
<hr>

In [52]:
road_type_counts = df['Road_Type'].value_counts()
road_type_counts

Road_Type
Single carriageway    496663
Dual carriageway       99424
Roundabout             43992
One way street         13559
Slip road               7041
Name: count, dtype: int64

<h3>INSIGHT: Single carriageways record the most accidents with 496663, as they carry high volumes of everyday traffic but lack the safety measures found on motorways and dual carriageways.</h3><hr>

<h1>Question8: How does weather condition affect accident frequency?</h1>
<hr>

In [54]:
weather_counts = df['Weather_Conditions'].value_counts()
weather_counts

Weather_Conditions
Fine no high winds       535013
Raining no high winds     79696
Other                     17150
Raining + high winds       9615
Fine + high winds          8554
Snowing no high winds      6238
Fog or mist                3528
Snowing + high winds        885
Name: count, dtype: int64

<h3>INSIGHT: Most accidents happen in fine weather conditions, not during storms or fog. This shows that risk is often happened during traffic exposure more cars on the road in good weather rather than just dangerous weather itself.</h3>

<h1>Question9: How does road surface condition affect accident severity?</h1>
<hr>

In [56]:
severity_by_surface = df.groupby(['Road_Surface_Conditions','Accident_Severity']).size().unstack(fill_value=0)
severity_by_surface

Accident_Severity        Fatal  Serious  Slight
Road_Surface_Conditions                        
Dry                       5790    61708  381049
Flood over 3cm. deep        23      152     842
Frost or ice               193     2007   16317
Snow                        35      565    5290
Wet or damp               2620    23785  160303

<h3>INSIGHT: Dry roads show the most accidents overall, since most driving happens in those conditions.</h3>

<h1>Question10: What is the distribution of number of vehicles involved in accidents?</h1>
<hr>

In [58]:
vehicle_counts = df['Number_of_Vehicles'].value_counts().sort_index()
vehicle_counts

Number_of_Vehicles
1     200787
2     391995
3      52806
4      11300
5       2464
6        758
7        296
8        147
9         57
10        27
11        12
12         7
13         6
14         9
15         1
16         4
19         1
28         1
32         1
Name: count, dtype: int64

<h3>INSIGHT: Most accidents involve just two vehicles, reflecting common collisions like rear-ends or side impacts. Single-vehicle crashes occur less often but are more likely to be severe.</h3>

<h1>Question11: What is the average number of casualties per accident by severity?</h1>
<hr>

In [59]:
avg_casualties = df.groupby('Accident_Severity')['Number_of_Casualties'].mean()
avg_casualties

Accident_Severity
Fatal      1.903129
Serious    1.467280
Slight     1.331402
Name: Number_of_Casualties, dtype: float64

<h3>INSIGHT: Fatal and serious accidents have a higher average number of casualties than slight accidents. This shows that when collisions are more severe, they often impact multiple people.</h3>

<h1>Question12: How does the number of vehicles involved impact accident severity?</h1><hr>

In [61]:
severity_by_vehicles = df.groupby('Number_of_Vehicles')['Accident_Severity'].value_counts(normalize=True).unstack(fill_value=0)
severity_by_vehicles

Accident_Severity      Fatal   Serious    Slight
Number_of_Vehicles                              
1                   0.019349  0.193937  0.786714
2                   0.008845  0.106068  0.885088
3                   0.017044  0.109988  0.872969
4                   0.024071  0.118584  0.857345
5                   0.028409  0.132305  0.839286
6                   0.036939  0.159631  0.803430
7                   0.057432  0.212838  0.729730
8                   0.081633  0.170068  0.748299
9                   0.017544  0.087719  0.894737
10                  0.074074  0.185185  0.740741
11                  0.166667  0.000000  0.833333
12                  0.000000  0.285714  0.714286
13                  0.000000  0.333333  0.666667
14                  0.333333  0.000000  0.666667
15                  0.000000  0.000000  1.000000
16                  0.250000  0.250000  0.500000
19                  0.000000  1.000000  0.000000
28                  1.000000  0.000000  0.000000
32                  0.000000  0.000000  1.000000

<h3>INSIGHT: Single-vehicle crashes are more likely to be fatal compared to multi-vehicle accidents.Because of fast speed, while multiple vehicle are slight because of traffic.</h3><hr>

<h1>Question13: Do accidents increase during holidays (Christmas, New Year)?</h1><hr>

In [63]:
df['Day'] = df['Accident Date'].dt.day
df['Holiday_Period'] = df['Accident Date'].apply(
    lambda x: 'Holiday' if (x.month == 12 and x.day >= 24) or (x.month == 12 and x.day <= 31) or (x.month == 1 and x.day <= 1) else 'Non-Holiday'
)
holiday_counts = df['Holiday_Period'].value_counts()
holiday_counts

Holiday_Period
Non-Holiday    637031
Holiday         23648
Name: count, dtype: int64

<h3>INSIGHT: There is a noticeable spike in accidents during the non-holiday period. Since people go to work and school which cause more riders/vehicle to actively travel</h3><hr>

<h1>Question14: What are the most common vehicle types involved in accidents?</h1><hr>

In [64]:
vehicle_type_counts = df['Vehicle_Type'].value_counts().head(10)
vehicle_type_counts

Vehicle_Type
Car                                      497992
Van / Goods 3.5 tonnes mgw or under       34160
Bus or coach (17 or more pass seats)      25878
Motorcycle over 500cc                     25657
Goods 7.5 tonnes mgw and over             17307
Motorcycle 125cc and under                15269
Taxi/Private hire car                     13294
Motorcycle over 125cc and up to 500cc      7656
Motorcycle 50cc and under                  7603
Goods over 3.5t. and under 7.5t            6096
Name: count, dtype: int64

<h3>INSIGHT: Cars make up the bulk of vehicles involved in accidents, simply because they dominate traffic volume. Motorcycles and vans also appear prominently, while heavy vehicles and buses are much less frequent.</h3><hr>

<h1>Question15: Which districts record the most accidents?</h1><hr>

In [66]:
district_counts = df['District Area'].value_counts().head(10)
district_counts

District Area
Birmingham          13491
Leeds                8898
Manchester           6720
Bradford             6212
Sheffield            5710
Westminster          5706
Liverpool            5587
Glasgow City         4942
Bristol, City of     4819
Kirklees             4690
Name: count, dtype: int64

<h3>INSIGHT: The busiest districts which is Birmingham record the highest accident numbers with 13491. High population density, traffic congestion, and more complex road networks drive this trend.</h3><hr>

<h1>Question16: Are multi-vehicle crashes deadlier than single-vehicle ones?</h1><hr>

In [68]:
multi_vs_single = df.groupby('Number_of_Vehicles')['Accident_Severity'].value_counts(normalize=True).unstack(fill_value=0)
multi_vs_single

Accident_Severity      Fatal   Serious    Slight
Number_of_Vehicles                              
1                   0.019349  0.193937  0.786714
2                   0.008845  0.106068  0.885088
3                   0.017044  0.109988  0.872969
4                   0.024071  0.118584  0.857345
5                   0.028409  0.132305  0.839286
6                   0.036939  0.159631  0.803430
7                   0.057432  0.212838  0.729730
8                   0.081633  0.170068  0.748299
9                   0.017544  0.087719  0.894737
10                  0.074074  0.185185  0.740741
11                  0.166667  0.000000  0.833333
12                  0.000000  0.285714  0.714286
13                  0.000000  0.333333  0.666667
14                  0.333333  0.000000  0.666667
15                  0.000000  0.000000  1.000000
16                  0.250000  0.250000  0.500000
19                  0.000000  1.000000  0.000000
28                  1.000000  0.000000  0.000000
32                  0.000000  0.000000  1.000000

<h3>INSIGHT: Single-vehicle accidents are more likely to be fatal, often involving loss of control at speed or collisions with fixed objects. Multi-vehicle crashes, while more common, usually result in slight injuries.</h3><hr>

<h1>Question17: Which combination of light and weather conditions is most dangerous?</h1><hr>

In [70]:
combo = df.groupby(['Light_Conditions','Weather_Conditions'])['Accident_Severity'].value_counts().unstack(fill_value=0)
combo


Accident_Severity                                  Fatal  Serious  Slight
Light_Conditions            Weather_Conditions                           
Darkness - lighting unknown Fine + high winds          3        7      49
                            Fine no high winds        57      662    4614
                            Fog or mist                0        6      59
                            Other                      1       31     154
                            Raining + high winds       2       14      95
                            Raining no high winds      5       66     550
                            Snowing + high winds       0        0      15
                            Snowing no high winds      0        8      86
Darkness - lights lit       Fine + high winds         44      311    1497
                            Fine no high winds      1482    14532   77944
                            Fog or mist               20      128     773
                            Other                     38      552    4213
                            Raining + high winds      44      426    2596
                            Raining no high winds    226     2996   19444
                            Snowing + high winds       1       32     208
                            Snowing no high winds      5      153    1670
Darkness - lights unlit     Fine + high winds          2        9      40
                            Fine no high winds        35      246    1464
                            Fog or mist                1        4      32
                            Other                      0       21     122
                            Raining + high winds       2       11      60
                            Raining no high winds      5       63     391
                            Snowing + high winds       0        1       4
                            Snowing no high winds      0        5      25
Darkness - no lighting      Fine + high winds         33      156     607
                            Fine no high winds      1250     5177   18824
                            Fog or mist               29      138     695
                            Other                     43      250    1622
                            Raining + high winds      49      238    1136
                            Raining no high winds    196     1082    4930
                            Snowing + high winds       1       26     144
                            Snowing no high winds     11      107     693
Daylight                    Fine + high winds         93      762    4941
                            Fine no high winds      4383    52668  351675
                            Fog or mist               32      207    1404
                            Other                     83      947    9073
                            Raining + high winds      48      572    4322
                            Raining no high winds    416     5261   44065
                            Snowing + high winds       1       50     402
                            Snowing no high winds     20      292    3163

<h3>INSIGHT: The most common combination is daylight with fine weather, simply because traffic is busiest then and riders tends to get in a high speed.</h3><hr>

<h1>Question18: What’s the trend of accidents over the years?</h1><hr>

In [71]:
trend = df['Accident Date'].dt.year.value_counts().sort_index()
trend

Accident Date
2019.0    71867
2020.0    70163
2021.0    66172
2022.0    56805
Name: count, dtype: int64

<h3>INSIGHT: Accident numbers show fluctuations over time, with some years trending downward due to better safety regulations, vehicle technology, and enforcement. Spikes may coincide with higher traffic volumes or seasonal effects.</h3><hr>

<h1>Question19: Are fatal accidents more common in rural coordinates (low density areas)?</h1><hr>

In [72]:
fatal_coords = df[df['Accident_Severity'] == 'Fatal'][['Latitude','Longitude']]
fatal_coords.describe()

Latitude    Longitude
count  8661.000000  8661.000000
mean     52.696000    -1.546123
std       1.538523     1.450183
min      50.083471    -6.430902
25%      51.517812    -2.522856
50%      52.436049    -1.499085
75%      53.501361    -0.362033
max      60.321208     1.751490

<h3>INSIGHT: Fatal crashes are more dispersed, often outside city clusters. Sparse coordinates reflect rural roads where higher speeds and slower emergency response matter.</h3><hr>

<h1>Question20: Do accident frequencies correlate with population-dense regions (lat/long clustering)?</h1><hr>

In [81]:
corr = df['Latitude'].corr(df['Number_of_Casualties'])
corr

np.float64(0.03220068662590638)

<h3>INSIGHT: There is a correlation between Latitude and Number of casualties. Towns and cities are more crowded, accidents usually involve more people. Busy roads mean that when a crash happens, it’s more likely to hurt multiple people at once.</h3>